extmod/modssl_mbedtls.c

```diff
@@ -63,6 +74,21 @@ typedef struct _mp_obj_ssl_socket_t {
     int last_error; // The last error code, if any
 } mp_obj_ssl_socket_t;
 
+
+typedef struct _mp_obj_ssl_context_t {
+    mp_obj_base_t base;
+    mp_obj_t sock;
+    mbedtls_entropy_context entropy;
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_ssl_context ssl;
+    mbedtls_ssl_config conf;
+    mbedtls_x509_crt cacert;
+    mbedtls_x509_crt cert;
+    mbedtls_pk_context pkey;
```

Comment:
All this context/state has been added here (in the SSLContext object) but it also remains in the SSLSocket object.  I don't think that's the correct approach because (1) it uses additional memory, to duplicate the state in each SSLSocket, (2) it means that a multiple SSLSockets can't share state from the same SSLContext.

I think the only thing that should be in the SSLSocket struct is `mbedtls_ssl_context` (the naming is poor, but `mbedtls_ssl_config` maps to SSLContext and `mbedtls_ssl_context` maps to SSLSocket).